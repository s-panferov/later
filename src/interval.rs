use chrono::{DateTime, Duration, Utc};
use std::fmt::Debug;

#[derive(Debug, Clone)]
pub struct Interval {
	pub from: DateTime<Utc>,
	pub to: Option<DateTime<Utc>>,
}

impl Interval {
	pub fn from_now() -> Self {
		Interval {
			from: Utc::now(),
			to: None,
		}
	}

	pub fn from(from: DateTime<Utc>) -> Self {
		Interval { from, to: None }
	}
}

/// Something that resolves to the range of time
pub trait AsIntervals: Debug {
	/// Get the estimate of the interval
	fn duration_hint(&self) -> Duration;

	/// Get an interator to resolve intervals
	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>>;
}
