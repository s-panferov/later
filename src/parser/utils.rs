use nom::{
	branch::alt,
	bytes::complete::tag,
	character::complete::{space0, space1},
	combinator::opt,
	sequence::tuple,
};

use super::error::ParseResult;

pub fn parse_chain<T, F>(input: &str, parser: F) -> ParseResult<Vec<T>>
where
	F: Fn(&str) -> ParseResult<T>,
{
	let mut res = vec![];
	let sep = opt(alt((
		tuple((space0, tag(","), space0)),
		tuple((space1, tag("and"), space1)),
	)));

	let (mut input, value) = parser(input)?;
	res.push(value);

	loop {
		let (i, s) = sep(input)?;
		match s {
			Some(_) => {
				let (i, value) = parser(i)?;
				res.push(value);
				input = i
			}
			None => break,
		}
	}

	Ok((input, res))
}
