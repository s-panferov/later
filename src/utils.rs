use chrono::{DateTime, Datelike, NaiveDate, Timelike, Utc};

pub enum Of {
	Day,
	Month,
	Year,
}

pub fn end_of(of: Of, date: &DateTime<Utc>) -> DateTime<Utc> {
	match of {
		Of::Year => end_of(Of::Month, &date.with_month(12).unwrap()),
		Of::Month => end_of(
			Of::Day,
			&date
				.with_day(days_in_month(date.year(), date.month()))
				.unwrap(),
		),
		Of::Day => date
			.with_hour(23)
			.unwrap()
			.with_minute(59)
			.unwrap()
			.with_second(59)
			.unwrap()
			.with_nanosecond(999999999)
			.unwrap(),
	}
}

pub fn start_of(of: Of, date: &DateTime<Utc>) -> DateTime<Utc> {
	match of {
		Of::Year => start_of(Of::Month, &date.with_month(1).unwrap()),
		Of::Month => start_of(Of::Day, &date.with_day(1).unwrap()),
		Of::Day => date
			.with_hour(0)
			.unwrap()
			.with_minute(0)
			.unwrap()
			.with_second(0)
			.unwrap()
			.with_nanosecond(0)
			.unwrap(),
	}
}

pub fn days_in_month(year: i32, m: u32) -> u32 {
	if m == 12 {
		NaiveDate::from_ymd(year + 1, 1, 1)
	} else {
		NaiveDate::from_ymd(year, m + 1, 1)
	}
	.signed_duration_since(NaiveDate::from_ymd(year, m, 1))
	.num_days() as u32
}
