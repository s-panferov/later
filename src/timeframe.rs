use chrono::{Duration, NaiveTime, Timelike};

use crate::interval::{Interval, Timeline};
use crate::utils::{end_of, start_of, Of};

#[derive(Debug, Clone)]
struct TimeFrame {
	pub from: NaiveTime,
	pub to: NaiveTime,
}

struct TimeFrameIterator {
	frame: TimeFrame,
	interval: Interval,
	completed: bool,
}

impl Timeline for TimeFrame {
	fn duration_hint(&self) -> Duration {
		if self.to > self.from {
			self.to - self.from
		} else {
			Duration::days(1)
				- self
					.from
					.signed_duration_since(NaiveTime::from_hms(0, 0, 0))
				+ self.to.signed_duration_since(NaiveTime::from_hms(0, 0, 0))
		}
	}

	fn iter_within(
		&self,
		interval: Interval,
	) -> Box<dyn Iterator<Item = Interval>> {
		Box::new(TimeFrameIterator {
			frame: self.clone(),
			interval,
			completed: false,
		})
	}
}

impl Iterator for TimeFrameIterator {
	type Item = Interval;

	fn next(&mut self) -> Option<Interval> {
		if self.completed {
			return None;
		}

		let mut from = self.interval.from.clone();
		if from.time() > self.frame.from {
			// Jump to the next day
			from = start_of(
				Of::Day,
				&(end_of(Of::Day, &from) + Duration::seconds(1)),
			);
		}

		from = from
			.with_hour(self.frame.from.hour())
			.and_then(|time| time.with_minute(self.frame.from.minute()))
			.and_then(|time| time.with_second(self.frame.from.second()))
			.unwrap();

		if self.interval.to.is_some()
			&& from > *self.interval.to.as_ref().unwrap()
		{
			self.completed = true;
			return None;
		}

		let mut to = from;
		if to.time() > self.frame.to {
			// Jump to the next day
			to = start_of(
				Of::Day,
				&(end_of(Of::Day, &from) + Duration::seconds(1)),
			);
		}

		to = to
			.with_hour(self.frame.to.hour())
			.and_then(|time| time.with_minute(self.frame.to.minute()))
      .and_then(|time| time.with_second(self.frame.to.second()))
			.unwrap();

		if self.interval.to.is_some()
			&& to > *self.interval.to.as_ref().unwrap()
		{
			self.completed = true;
			return None;
		}

		self.interval.from = to + Duration::seconds(1);

		Some(Interval { from, to: Some(to) })
	}
}

#[cfg(test)]
mod tests {
	use super::*;
	use crate::interval::{Interval, Timeline};
	use insta::assert_debug_snapshot_matches;

	#[test]
	fn from_23_00_to_1_00() {
		let frame = TimeFrame {
			from: NaiveTime::from_hms(23, 0, 0),
			to: NaiveTime::from_hms(1, 0, 0),
		};

		assert_debug_snapshot_matches!(
			"from_23_00_to_1_00",
			frame
				.iter_within(Interval::from(
					"2019-10-31T00:00:00Z".parse().unwrap(),
				))
				.take(10)
				.collect::<Vec<_>>()
		);
	}
}
