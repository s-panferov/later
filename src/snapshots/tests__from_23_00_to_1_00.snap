---
created: "2019-09-05T05:08:22.943204Z"
creator: insta@0.10.1
source: src/timeframe.rs
expression: "frame.iter_within(Interval::from(\"2019-10-31T00:00:00Z\".parse().unwrap())).take(10).collect::<Vec<_>>()"
---
[
    Interval {
        from: 2019-10-31T23:00:00Z,
        to: Some(
            2019-11-01T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-01T23:00:00Z,
        to: Some(
            2019-11-02T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-02T23:00:00Z,
        to: Some(
            2019-11-03T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-03T23:00:00Z,
        to: Some(
            2019-11-04T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-04T23:00:00Z,
        to: Some(
            2019-11-05T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-05T23:00:00Z,
        to: Some(
            2019-11-06T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-06T23:00:00Z,
        to: Some(
            2019-11-07T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-07T23:00:00Z,
        to: Some(
            2019-11-08T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-08T23:00:00Z,
        to: Some(
            2019-11-09T01:00:00Z,
        ),
    },
    Interval {
        from: 2019-11-09T23:00:00Z,
        to: Some(
            2019-11-10T01:00:00Z,
        ),
    },
]
